### PR TITLE
Detector can now receive a buffer as parameter for both cfg and weights

### PR DESCRIPTION
**Pull request by grandec**
This was a feature requested in #6019 .
As the constructors were the same, I had to make another one with another signature.

### CODE REVIEW

**Comment by imaami** [comment]
Why are so many functions duplicated? Is reading weights from memory somehow different from first reading them into memory from a file? Isn't a memory buffer-based weights importing function simply the current one but with the file reading part removed?

**Comment by grandec** [comment]
> Why are so many functions duplicated? Is reading weights from memory somehow different from first reading them into memory from a file? Isn't a memory buffer-based weights importing function simply the current one but with the file reading part removed?

I "duplicated" the functions so that this new functionality do not interfere with all the working code that currently is running. 
So any bug that I might have put there, do not propagate.
But I see your point, if we create a "merged", there would not be a reason to do all this duplication.

**Comment by imaami** [comment]
> > Why are so many functions duplicated? Is reading weights from memory somehow different from first reading them into memory from a file? Isn't a memory buffer-based weights importing function simply the current one but with the file reading part removed?
> 
> I "duplicated" the functions so that this new functionality do not interfere with all the working code that currently is running.
> So any bug that I might have put there, do not propagate.
> But I see your point, if we create a "merged" one there would not be a reason to do all this duplication.

Yeah, code re-use is usually a good thing. Copypasting functions quickly leads to a maintainability nightmare; fixing a bug in the shared logic requires modifying every copy, keeping track of the duplicated code makes reading the source unnecessarily taxing mentally, the compiled binary is significantly larger than it needs to be, etc.

And of course the commits you made here can't actually be compared against the existing functions reasonably. If the change is simply long blocks of added lines which are mostly copies, then you can't do `git show HEAD` and see the changed parts. You have to save the old functions in one file, save the new functions in another file, and diff those...

**Comment by grandec** [comment]
In this last commit I developed some of the tips that you guys gave here and in #6019.
The constructor is changed and most of the duplicated code is now gone.

**Comment by grandec** [comment]
In this latest commit I changed the new constructors to be
`Detector(const std::string& cfg_string, const char* weights, size_t size, int gpu_id);`
and
`Detector(const std::string& cfg_string, const std::vector<char>& weights, int gpu_id = 0): Detector(cfg_string, weights.data(), weights.size(), gpu_id) {};`

as suggested by @imaami

**Comment by imaami** [comment]
Hi @grandec , sorry for taking this long to reply! I've a couple of change requests.

1. Quoting from [\#6019](https://github.com/AlexeyAB/darknet/issues/6019#issuecomment-650440273):

    > > @Kaimbelic Just to confirm: you only need the weights data to be in memory, not the cfg text as well? Do you still read the cfg normally as a text file from disk?
    > 
    > Yes, just the weights.

    I.e., the first argument shouldn't be the cfg file's contents, it should just be the cfg filename like in the existing constructor. I'd still change it from `std::string cfg_filename` to `std::string const &cfg_filename` because it's not meant to be modified by the constructor. (For this same reason I also think the existing constructor uses the wrong type for `cfg_filename` and `weight_filename`, but possibly changing those is out of scope here.)

2. The type for the weights data pointer should not be `const char *`, it should be `const std::uint8_t *` (or `std::uint8_t const *` if you prefer, it's the same type).

    The weights data is raw binary bytes, not a null-terminated character string. Although `const char *` _can_ technically be used to address raw binary bytes, any C/C++ coder will automatically assume it's a C string pointer that can be passed to e.g. `strlen()` and the like. It's not good design.
 
    While I think the primary issue with `const char *` is the fact that it's misleading from a human point of view, there's also another, slightly more hypothetical problem: a `char` is neither guaranteed to be 8 bits, nor is it explicitly signed or unsigned. The width and sign of a `char` is implementation-defined in C. I know that _in practice_ a `char` is 8 bits on any device that Darknet runs on, but it's nevertheless wrong to assume that because, well, language specifications are what they are.

    As we're unambiguously passing a read-only buffer of generic binary bytes to the constructor, the correct types are `std::uint8_t const *` and `std::vector<std::uint8_t> const &`.

**Comment by grandec** [comment]
> Hi @grandec , sorry for taking this long to reply! I've a couple of change requests.
> 
> 1. Quoting from [#6019](https://github.com/AlexeyAB/darknet/issues/6019#issuecomment-650440273):
>    > > @Kaimbelic Just to confirm: you only need the weights data to be in memory, not the cfg text as well? Do you still read the cfg normally as a text file from disk?
>    > 
>    > 
>    > Yes, just the weights.
>    
>    
>    I.e., the first argument shouldn't be the cfg file's contents, it should just be the cfg filename like in the existing constructor. I'd still change it from `std::string cfg_filename` to `std::string const &cfg_filename` because it's not meant to be modified by the constructor. (For this same reason I also think the existing constructor uses the wrong type for `cfg_filename` and `weight_filename`, but possibly changing those is out of scope here.)
>    This means that the config parsing code in your PR is not needed, we just use the existing config file parsing function.
> 2. The type for the weights data pointer should not be `const char *`, it should be `const std::uint8_t *` (or `std::uint8_t const *` if you prefer, it's the same type).
>    The weights data is raw binary bytes, not a null-terminated character string. Although `const char *` _can_ technically be used to address raw binary bytes, any C/C++ coder will automatically assume it's a C string pointer that can be passed to e.g. `strlen()` and the like. It's not good design.
>    While I think the primary issue with `const char *` is the fact that it's misleading from a human point of view, there's also another, slightly more hypothetical problem: a `char` is neither guaranteed to be 8 bits, nor is it explicitly signed or unsigned. The width and sign of a `char` is implementation-defined in C. I know that _in practice_ a `char` is 8 bits on any device that Darknet runs on, but it's nevertheless wrong to assume that because, well, language specifications are what they are.
>    As we're unambiguously passing a read-only buffer of generic binary bytes to the constructor, the correct types are `std::uint8_t const *` and `std::vector<std::uint8_t> const &`.
> 
> **TL;DR:**
> 
> ```
> Detector(std::string const &cfg_filename, std::uint8_t const *weights, std::size_t size, int gpu_id);
> 
> Detector(std::string const &cfg_filename, std::vector<std::uint8_t> &weights, int gpu_id = 0)
>         : Detector(cfg_string, weights.data(), weights.size(), gpu_id) { }
> ```

1) About the first parameter being a filename, I currently am working on a project that it is necessary to pass both cfg and weights using a buffer. So at least for me, if I change the first parameter to a filename, I would have to open another Issue to solve it (Should I open another one?)

2) About changing to const and uint8_t, I see no problem in making the changes,  making sure that it is a stream of bytes with uint8_t seems nice.

**Comment by grandec** [comment]
In this last commit, 2 of the checks were not sucessful and I don't know why :(

File: src/parser.c [hunk 1]
```diff
@@ -1811,20 +1817,55 @@ network parse_network_cfg_custom(char *filename, int batch, int time_steps)
...
+    data_size = ftell( fp );
+    rewind( fp );
+
+    /* allocate memory for entire content */
+    data = (char*) xmalloc(data_size + 1);
+
+    /* copy the file into the buffer */
+    if( fread( data , data_size, 1 , fp) != 1)
+    {
+        error("Copy file to buffer failed");
```
Comment (cenit) [inline src/parser.c:1839]:
```suggestion
        error("Copy file to buffer failed", DARKNET_LOC);
```

File: src/parser.c [hunk 2]
```diff
@@ -1811,20 +1817,55 @@ network parse_network_cfg_custom(char *filename, int batch, int time_steps)
...
+    return net;
+}
+
+list *read_cfg_mem(const char* str)
 {
-    FILE *file = fopen(filename, "r");
-    if(file == 0) file_error(filename);
-    char *line;
-    int nu = 0;
+    if(str == NULL) error("string is NULL!");
```
Comment (cenit) [inline src/parser.c:1853]:
```suggestion
    if(str == NULL) error("string is NULL!", DARKNET_LOC);
```

File: src/parser.c [hunk 3]
```diff
@@ -1838,13 +1879,47 @@ list *read_cfg(char *filename)
...
+    data_size = ftell( fp );
+    rewind( fp );
+
+    /* allocate memory for entire content */
+    data = (char*) xmalloc(data_size + 1);
+
+    /* copy the file into the buffer */
+    if( fread( data , data_size, 1 , fp) != 1)
+    {
+        error("Copy file to buffer failed");
```
Comment (cenit) [inline src/parser.c:1913]:
```suggestion
        error("Copy file to buffer failed", DARKNET_LOC);
```

File: src/parser.c [hunk 4]
```diff
@@ -2240,49 +2306,33 @@ void load_shortcut_weights(layer l, FILE *fp)
...
 #ifdef GPU
     if(net->gpu_index >= 0){
         cuda_set_device(net->gpu_index);
     }
 #endif
-    fprintf(stderr, "Loading weights from %s...", filename);
-    fflush(stdout);
-    FILE *fp = fopen(filename, "rb");
-    if(!fp) file_error(filename);
+    if(!data) error("Weights data is NULL");
```
Comment (cenit) [inline src/parser.c:2316]:
```suggestion
    if(!data) error("Weights data is NULL", DARKNET_LOC);
```

File: src/parser.c [hunk 5]
```diff
@@ -2294,87 +2344,142 @@ void load_weights_upto(network *net, char *filename, int cutoff)
...
+    data_size = ftell( fp );
+    rewind( fp );
+
+    /* allocate memory for entire content */
+    data = (char*) xmalloc(data_size);
+
+    /* copy the file into the buffer */
+    if( fread( data , data_size, 1 , fp) != 1)
+    {
+        error("Copy file to buffer failed");
```
Comment (cenit) [inline src/parser.c:2448]:
```suggestion
        error("Copy file to buffer failed", DARKNET_LOC);
```

File: src/parser.c [hunk 6]
```diff
@@ -2294,87 +2344,142 @@ void load_weights_upto(network *net, char *filename, int cutoff)
...
         if(l.type == CONVOLUTIONAL && l.share_layer == NULL){
-            load_convolutional_weights(l, fp);
+            load_convolutional_weights(l, data, data_size, &data_index);
         }
         if (l.type == SHORTCUT && l.nweights > 0) {
-            load_shortcut_weights(l, fp);
+            load_shortcut_weights(l, data, data_size, &data_index);
         }
         if (l.type == IMPLICIT) {
             load_implicit_weights(l, fp);
```
Comment (cenit) [inline src/parser.c:2353]:
```suggestion
            load_implicit_weights(l, data, data_size, &data_index);
```

File: src/parser.c [hunk 7]
```diff
@@ -2240,49 +2306,33 @@ void load_shortcut_weights(layer l, FILE *fp)
 #endif
 }
```
Comment (cenit) [inline src/parser.c:2308]:
```suggestion
void load_implicit_weights(layer l, const char* data, int data_size, int* data_index)
{
    int num = l.nweights;
    int read_bytes;
    read_bytes = mread(l.weights, sizeof(float), num, data, data_size, data_index);
    if (read_bytes > 0 && read_bytes < num)
        printf("\n Warning: Unexpected end of wights-file! l.weights - l.index = %d \n", l.index);
        // for (int i = 0; i < l.nweights; ++i) printf(" %f, ", l.weights[i]);
        // printf(" read_bytes = %d \n\n", read_bytes);
#ifdef GPU
    if (gpu_index >= 0) {
        push_implicit_layer(l);
    }
#endif
}
```

File: src/parser.c [hunk 8]
```diff
@@ -2239,50 +2305,48 @@ void load_shortcut_weights(layer l, FILE *fp)
...
 
-    int32_t major;
-    int32_t minor;
-    int32_t revision;
-    fread(&major, sizeof(int32_t), 1, fp);
-    fread(&minor, sizeof(int32_t), 1, fp);
-    fread(&revision, sizeof(int32_t), 1, fp);
+    int major;
+    int minor;
+    int revision;
```
Comment (cenit) [inline src/parser.c:2334]:
```suggestion
    int32_t major;
    int32_t minor;
    int32_t revision;
```

File: src/parser.c [hunk 9]
```diff
@@ -2239,50 +2305,48 @@ void load_shortcut_weights(layer l, FILE *fp)
     }
 #endif
 }
-
-void load_implicit_weights(layer l, FILE *fp)
+void load_implicit_weights(layer l, const char* data, int data_size, int* data_index)
```
Comment (cenit) [inline src/parser.c:2308]:
```suggestion

void load_implicit_weights(layer l, const char* data, int data_size, int* data_index)
```

File: src/parser.c [hunk 10]
```diff
@@ -2239,50 +2305,48 @@ void load_shortcut_weights(layer l, FILE *fp)
...
+        // printf(" read_bytes = %d \n\n", read_bytes);
 #ifdef GPU
     if (gpu_index >= 0) {
         push_implicit_layer(l);
     }
 #endif
 }
-
-void load_weights_upto(network *net, char *filename, int cutoff)
+void load_weights_upto_mem(network *net, const char *data, int data_size, int cutoff)
```
Comment (cenit) [inline src/parser.c:2323]:
```suggestion

void load_weights_upto_mem(network *net, const char *data, int data_size, int cutoff)
```

File: src/parser.c [hunk 11]
```diff
@@ -2240,49 +2306,49 @@ void load_shortcut_weights(layer l, FILE *fp)
...
     int32_t minor;
     int32_t revision;
-    fread(&major, sizeof(int32_t), 1, fp);
-    fread(&minor, sizeof(int32_t), 1, fp);
-    fread(&revision, sizeof(int32_t), 1, fp);
+    int data_index = 0;
+
+    mread(&major, sizeof(int), 1, data, data_size, &data_index);
+    mread(&minor, sizeof(int), 1, data, data_size, &data_index);
+    mread(&revision, sizeof(int), 1, data, data_size, &data_index);
```
Comment (cenit) [inline src/parser.c:2341]:
```suggestion
    mread(&major, sizeof(int32_t), 1, data, data_size, &data_index);
    mread(&minor, sizeof(int32_t), 1, data, data_size, &data_index);
    mread(&revision, sizeof(int32_t), 1, data, data_size, &data_index);
```
